sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/container/ThroughputContainerController.java

```diff
@@ -0,0 +1,321 @@
+                this.updateMaxContainerThroughput(throughputResponse);
+                return Mono.empty();
+            })
+            .retryWhen(
+                // Throughput can be configured on database level or container level
+                // Retry at most 1 time so we can try on database and container both
+                RetrySpec.max(1).filter(throwable -> this.isOfferNotConfiguredException(throwable))
+            ).thenReturn(this);
+    }
+
+    private Mono<ThroughputResponse> resolveThroughputByResourceId(String resourceId) {
+        // TODO: figure out how this work for serveless account
+        // TODO: work item: https://github.com/Azure/azure-sdk-for-java/issues/18776
+        checkArgument(StringUtils.isNotEmpty(resourceId), "ResourceId can not be null or empty");
+        return this.client.queryOffers(resourceId, new CosmosQueryRequestOptions())
+            .single()
+            .flatMap(offerFeedResponse -> {
+                if (offerFeedResponse.getResults().isEmpty()) {
+                    return Mono.error(
+                        BridgeInternal.createCosmosException(NO_OFFER_EXCEPTION_STATUS_CODE, "No offers found for the resource " + resourceId));
+                }
+
+                return this.client.readOffer(offerFeedResponse.getResults().get(0).getSelfLink()).single();
+            })
+            .onErrorResume(throwable -> {
+                System.out.println(throwable.getCause());
+                return Mono.error(throwable);
+            })
+            .map(ModelBridgeInternal::createThroughputRespose);
+    }
+
+    private void setDefaultGroupController() {
+        List<ThroughputGroupControllerBase> defaultGroupControllers =
+            this.groupControllers.values().stream().filter(ThroughputGroupControllerBase::isUseByDefault).collect(Collectors.toList());
+
+        if (defaultGroupControllers.size() > 1) {
+            // We should never reach here since we do config validation to make sure only one default group configured
+            // when enable throughput control
+
+            throw new IllegalArgumentException("There should only be one default throughput control group");
+        }
+        if (defaultGroupControllers.size() == 1) {
+            this.defaultGroupController = defaultGroupControllers.get(0);
+        }
+    }
+
+    private void updateMaxContainerThroughput(ThroughputResponse throughputResponse) {
+        checkNotNull(throughputResponse, "Throughput response can not be null");
+
+        ThroughputProperties throughputProperties = throughputResponse.getProperties();
+        this.maxContainerThroughput.set(
+            Math.max(throughputProperties.getAutoscaleMaxThroughput(), throughputProperties.getManualThroughput()));
+    }
+
+    private boolean isOfferNotConfiguredException(Throwable throwable) {
+        checkNotNull(throwable, "Throwable should not be null");
+
+        CosmosException cosmosException = Utils.as(throwable, CosmosException.class);
```

Comment:
Please unwrap the exception here, like mentioned above in one of the comments.

Reply:
Ack, will add in the next iteration.